src/util/cpuid.cc

```diff
@@ -0,0 +1,218 @@
+  return cpu_family;
+}
+
+unsigned int CpuidCommand::GetCpuModel()
+{
+  unsigned int cpuid[4]{};
+  if (!CryptoPP::CpuId(1, 0, cpuid))
+    {
+      return 0;
+    }
+
+  unsigned int cpu_model = (cpuid[0] >> 4) & 0xf;
+  return cpu_model;
+}
+
+unsigned int CpuidCommand::GetCpuExtendedModel()
+{
+  unsigned int cpuid[4]{};
+  if (!CryptoPP::CpuId(1, 0, cpuid))
+    {
+      return 0;
+    }
+
+  unsigned int cpu_model = (cpuid[0] >> 16) & 0xf;
+  return cpu_model;
+}
+
+unsigned int CpuidCommand::GetCpuStepping()
+{
+  unsigned int cpuid[4]{};
+  if (!CryptoPP::CpuId(1, 0, cpuid))
+    {
+      return 0;
+    }
+
+  return cpuid[0] & 0xf;
+}
+#endif
+
+void CpuidCommand::PrintUsage(const std::string& name) const
+{
+  LOG(info) << "Syntax: " << name;
+}
+
+bool CpuidCommand::Impl(
+    const std::string& cmd_name,
+    const std::vector<std::string>& args)
+{
+#ifdef WITH_CRYPTOPP
+#if defined(__x86_64__) || defined(__i386__)
+  LOG(info) << "CPU Vendor: " << GetCpuVendor();
+  LOG(info) << "CPU Model Name: " << GetCpuModelName();
+  const int extended_val = 15;
+  const int extended_intel_val = 6;
+  int cpu_family = GetCpuFamily();
+  int cpu_model = GetCpuModel();
+  if (cpu_family == extended_val
+      || (cpu_family == extended_intel_val
+          && GetCpuVendor().compare("GenuineIntel") == 0))
+    {
+      cpu_model += (GetCpuExtendedModel() << 4);
+    }
+  if (cpu_family == extended_val)
+    {
+      cpu_family += GetCpuExtendedFamily();
+    }
+  LOG(info) << "CPU Family: " << cpu_family;
+  LOG(info) << "CPU Model: " << cpu_model;
+  LOG(info) << "CPU Stepping: " << GetCpuStepping();
+  LOG(info) << "CacheLineSize: " << CryptoPP::GetCacheLineSize();
+  LOG(info) << "Has SSE2: " << CryptoPP::HasSSE2();
+  LOG(info) << "Has SSSE3: " << CryptoPP::HasSSSE3();
+  LOG(info) << "Has SSE4.1: " << CryptoPP::HasSSE41();
+  LOG(info) << "Has SSE4.2: " << CryptoPP::HasSSE42();
+  LOG(info) << "Has AESNI: " << CryptoPP::HasAESNI();
+  LOG(info) << "Has CLMUL: " << CryptoPP::HasCLMUL();
+  LOG(info) << "Has SHA: " << CryptoPP::HasSHA();
+  LOG(info) << "Is P4: " << CryptoPP::IsP4();
+  LOG(info) << "Has RDRAND: " << CryptoPP::HasRDRAND();
+  LOG(info) << "Has RDSEED: " << CryptoPP::HasRDSEED();
+  LOG(info) << "Has PadlockRNG: " << CryptoPP::HasPadlockRNG();
+  LOG(info) << "Has PadlockACE: " << CryptoPP::HasPadlockACE();
+  LOG(info) << "Has PadlockACE2: " << CryptoPP::HasPadlockACE2();
+  LOG(info) << "Has PadlockPHE: " << CryptoPP::HasPadlockPHE();
+  LOG(info) << "Has PadlockPMM: " << CryptoPP::HasPadlockPMM();
+#elif defined(__arm__)
+  LOG(info) << "Has NEON: " << CryptoPP::HasNEON();
+  LOG(info) << "Has PMULL: " << CryptoPP::HasPMULL();
+  LOG(info) << "Has CRC32: " << CryptoPP::HasCRC32();
+  LOG(info) << "Has AES: " << CryptoPP::HasAES();
+  LOG(info) << "Has SHA1: " << CryptoPP::HasSHA1();
+  LOG(info) << "Has SHA2: " << CryptoPP::HasSHA2();
+#endif
+  return true;
+#endif
+  LOG(warning) << "cpuid requires Crypto++";
```

Comment:
Will there be a non-cryotopp implementation? If not, maybe we should reconsider how we treat pre-processing and simply leave out the command option all-together if cryptopp is not available.

Reply:
Note: doing so would cut down on the number of directives needed and keep things generally cleaner.